In [231]:
# Imports and pip installations (if needed)
import pandas as pd
import math
from sklearn.preprocessing import MinMaxScaler


# Part 1: Load the dataset

In [232]:
# Load the given datasets
cat_data = pd.read_csv("./chronic_kidney_disease_categorical.csv")
num_data = pd.read_csv("./chronic_kidney_disease_numerical.csv")

# Print the data
cat_data

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


In [233]:
num_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)


Answer: 
There are two datasets, one containing 14 columns including a unique id, containing categorical health information patients. The other dataset includes 12 columns including a unique id, which holds information on numerical health markers on patients. 

First dataset contains categorical markers including red blood cells (abnormal or normal), bacteria (notpresent or present), and hypertension (yes or no).

Second dataset contains numerical markers including age (integer), blood pressure (integer), and blood glucose random (integer).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [234]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
seen = []
total = 0
for i in range(len(num_data)):
    if num_data.iloc[i]["unique_id"] in seen:
        total = total + 1
    else:
        seen.append(num_data.iloc[i]["unique_id"])
print("total number of duplicate rows in numerical dataset: ", total)
    
# Drop these duplicate rows
seen = []
i = 0
while i < len(num_data):
    if num_data.iloc[i]["unique_id"] in seen:
        num_data = num_data.drop(i)
        num_data = num_data.reset_index(drop=True)
    else:
        seen.append(num_data.iloc[i]["unique_id"])
        i += 1

print(len(num_data))

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
seen = []
total = 0
for i in range(len(cat_data)):
    if cat_data.iloc[i]["unique_id"] in seen:
        total = total + 1
    else:
        seen.append(cat_data.iloc[i]["unique_id"])
print("total number of duplicate rows in categorical dataset: ", total)


seen = []
i = 0
while i < len(cat_data):
    if cat_data.iloc[i]["unique_id"] in seen:
        cat_data = cat_data.drop(i)
        cat_data = cat_data.reset_index(drop=True)
    else:
        seen.append(cat_data.iloc[i]["unique_id"])
        i += 1

print(len(cat_data))


total number of duplicate rows in numerical dataset:  30
390
total number of duplicate rows in categorical dataset:  30
400


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, I think this because there are 390 rows in the numerical dataset after dropping the duplicates and 400 in the categorical dataset after dropping duplicates. The rows that are missing don't get merged. 

In [235]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged = pd.merge(num_data, cat_data, on="unique_id")
#Print the combined dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5,0.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0,1.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4,4.0,2.0,abnormal,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,normal,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [236]:
# Calculate the percentage of rows that contain atleast one missing value

missing = 0
for i in range(len(merged)):
    for value in merged.iloc[i]:
        if pd.isna(value):
            missing += 1
            break


# Print %
print("Percentage of rows with missing values: ", (missing / len(merged)) * 100, "%")
# Drop these rows from the dataset
merged = merged.dropna()
# Print the Dataset
merged

Percentage of rows with missing values:  60.76923076923077 %


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
379,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,5800.0,6.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4,4.0,2.0,abnormal,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [237]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged = merged.sort_values(by="Target").reset_index(drop="True")

# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,normal,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,12100.0,3.7,2.0,0.0,abnormal,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,7000.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,7000.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,6200.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,6300.0,5.9,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [238]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
merged = pd.get_dummies(merged, columns=categorical_columns, prefix=categorical_columns)

# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,12100.0,3.7,2.0,0.0,True,False,True,False,False,True,True,False,False,True,False,True,True,False,False,True,True,False,False,True,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,7000.0,5.3,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,7000.0,5.2,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,6200.0,6.2,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,6300.0,5.9,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

If we have a single column with binary and exclusive value it would work if it could only take two values and it has to be one or the other. Instead of having a col for appet_good and appet_bad, we could just have the appet_good col and if its false we would know the appetite is bad becuase it's either one of the other. In the case of a sandwhich, a binary value would not work, as there are more than two values. If we tried to choose more values it still wouldnt work because there could be different kind of condiments at the same time.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [239]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for col in numerical_columns:
    mean = merged[col].mean()
    standard_dev = merged[col].std()
    lower_limit = mean - 3 * standard_dev
    upper_limit = mean + 3 * standard_dev

    merged = merged[(merged[col] >= lower_limit) & (merged[col] <= upper_limit)]
    

# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
6,546225,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,7000.0,5.3,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,7000.0,5.2,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,6200.0,6.2,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,6300.0,5.9,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [240]:
# Normalize the all Numerical Attributes in the dataset.
scaler = MinMaxScaler()
merged[numerical_columns] = scaler.fit_transform(merged[numerical_columns])
# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,0.257143,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,137148,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,0.057143,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
3,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,0.057143,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
4,532520,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,0.314286,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
6,546225,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,0.057143,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.206897,0.960396,0.935484,0.217742,0.657143,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
149,125587,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.689655,0.722772,0.806452,0.217742,0.628571,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
150,614376,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.620690,0.762376,0.612903,0.153226,0.914286,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
151,853426,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.724138,0.900990,0.580645,0.161290,0.828571,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column was important for cleaning the data, but won't be useful for predicting as it serves as an identifier and doesn't give us information on any health marker.

In [241]:
#Remove that column
merged = merged.drop(columns=["unique_id"])

# Print the dataset
merged = merged.reset_index(drop="True")
merged

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,0.257143,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,0.057143,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,0.057143,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
3,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,0.314286,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
4,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,0.057143,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.206897,0.960396,0.935484,0.217742,0.657143,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
129,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.689655,0.722772,0.806452,0.217742,0.628571,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
130,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.620690,0.762376,0.612903,0.153226,0.914286,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
131,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.724138,0.900990,0.580645,0.161290,0.828571,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [242]:
# Export the dataframe to a new csv file
merged.to_csv("cleaned_data.csv")
# Export the dataframe to a new json file
merged.to_json("cleaned_data.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [243]:
pd.set_option('display.max_columns', None)
print(merged.head(16))

         age    bp       bgr        bu        sc       sod       pot  \
0   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1   0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
2   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
3   0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
4   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
5   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
6   0.716216  0.50  1.000000  0.163399  0.111111  0.066667  0.206897   
7   0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
8   0.000000  0.00  0.103004  0.372549  0.074074  0.500000  0.689655   
9   0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.689655   
10  0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
11  0.878378  0.25  0.639485  0.470588  0.395062  0.433333  0.517241   
12  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.86

### Paste the markdown table here

| age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|--------|---------|----------|---------|---------|----------|----------|-----------|----------|-----------|-----------|---------|---------|---------------|------------|-------------|-----------|----------------|--------------|---------------|------------|--------|---------|-------|--------|--------|---------|-------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0.2027 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.000000 | 0.653226 | 0.257143 | 4.0 | 0.0 | False | True | True | False | False | True | False | True | True | False | True | False | True | False | True | False | True | False | False | True | True | False |
| 0.9054 | 1.00 | 0.965665 | 0.522876 | 0.641975 | 0.666667 | 0.000000 | 0.148515 | 0.225806 | 0.217742 | 0.057143 | 3.0 | 2.0 | True | False | True | False | False | True | True | False | False | True | False | True | False | True | False | True | True | False | True | False | True | False |
| 0.7432 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | 0.032258 | 0.395161 | 0.057143 | 2.0 | 0.0 | True | False | True | False | True | False | True | False | False | True | False | True | False | True | False | False | True | False | True | True | True | False |
| 0.7297 | 0.75 | 0.150215 | 0.281046 | 0.234568 | 0.533333 | 0.793103 | 0.336634 | 0.322581 | 0.500000 | 0.314286 | 2.0 | 0.0 | True | False | False | True | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | True | True | False |
| 0.5405 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.379310 | 0.207921 | 0.161290 | 0.830645 | 0.057143 | 1.0 | 0.0 | False | True | False | True | True | False | True | False | False | True | False | True | True | True | True | False | True | True | False | True | True | False |
| 0.6757 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.193548 | 0.169355 | 0.114286 | 2.0 | 0.0 | True | False | True | False | True | False | True | False | False | True | False | True | True | True | True | False | False | True | False | True | True | False |
| 0.7162 | 0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667 | 0.206897 | 0.267327 | 0.387097 | 0.532258 | 0.371429 | 1.0 | 0.0 | True | False | False | True | True | False | True | False | True | False | True | False | True | False | False | True | True | True | False | True | True | False |
| 0.7297 | 0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.034483 | 0.019802 | 0.064516 | 0.879032 | 0.000000 | 3.0 | 1.0 | False | True | True | False | False | True | True | False | True | False | True | False | True | False | False | True | True | True | True | False | True | False |
| 0.0000 | 0.00 | 0.103004 | 0.372549 | 0.074074 | 0.500000 | 0.689655 | 0.217822 | 0.225806 | 1.000000 | 0.514286 | 4.0 | 0.0 | True | False | True | False | True | False | True | False | True | False | True | False | True | False | False | True | True | True | True | False | True | False |
| 0.8784 | 0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333 | 0.689655 | 0.366337 | 0.387097 | 0.879032 | 0.371429 | 4.0 | 0.0 | False | True | False | True | True | False | True | False | True | False | True | False | True | False | False | True | False | True | True | False | True | False |
| 0.8514 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.344828 | 0.168317 | 0.161290 | 0.580645 | 0.085714 | 4.0 | 3.0 | False | True | True | False | False | True | True | False | True | False | True | False | True | False | True | False | False | False | True | True | True | False |
| 0.8784 | 0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433333 | 0.517241 | 0.267327 | 0.322581 | 0.104839 | 0.171429 | 3.0 |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

To mitigate data conversion hallucinations you could give the LLM as much context as possible. You could also give an example of what the first row should look like. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [244]:
# Code to rename all the columns in the dataset
# Define the column mapping
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class',
    'wbcc': 'white blood cell count corrected',  # Assuming a more descriptive name
    'rbcc': 'red blood cell count corrected',    # Assuming a more descriptive name
    'albumin': 'albumin',
    'sugar': 'sugar',
    'rbc_abnormal': 'red blood cells abnormal',
    'rbc_normal': 'red blood cells normal',
    'pc_abnormal': 'pus cell abnormal',
    'pc_normal': 'pus cell normal',
    'pcc_notpresent': 'pus cell clumps not present',
    'pcc_present': 'pus cell clumps present',
    'ba_notpresent': 'bacteria not present',
    'ba_present': 'bacteria present',
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'target chronic kidney disease',
    'Target_notckd': 'target not chronic kidney disease'
}

# Update the DataFrame's column names
merged.rename(columns=column_mapping, inplace=True)

merged


,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,white blood cell count corrected,red blood cell count corrected,albumin,sugar,red blood cells abnormal,red blood cells normal,pus cell abnormal,pus cell normal,pus cell clumps not present,pus cell clumps present,bacteria not present,bacteria present,hypertension no,hypertension yes,diabetes mellitus no,diabetes mellitus yes,coronary artery disease no,coronary artery disease yes,appetite good,appetite poor,pedal edema no,pedal edema yes,anemia no,anemia yes,target chronic kidney disease,target not chronic kidney disease
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,0.257143,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,0.057143,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,0.057143,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
3,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,0.314286,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
4,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,0.057143,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.206897,0.960396,0.935484,0.217742,0.657143,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
129,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.689655,0.722772,0.806452,0.217742,0.628571,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
130,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.620690,0.762376,0.612903,0.153226,0.914286,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
131,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.724138,0.900990,0.580645,0.161290,0.828571,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [245]:
# Converted SQL to Pandas code

# Assuming df is your DataFrame
# Example DataFrame
# df = pd.DataFrame({
#     'Target': [...],  # Replace with your actual data
#     # Other columns...
# })

# Group by 'Target' and count occurrences
result = merged.groupby('target chronic kidney disease').size().reset_index(name='count')

# Optionally, print the resulting DataFrame
print(result)


   target chronic kidney disease  count
0                          False    111
1                           True     22


*Describe what the above code does here*

It counts the occurences of each distinct value where TARGET is in the SQL query.